hypothesis-python/src/hypothesis/stateful.py

```diff
@@ -204,7 +208,11 @@ def run_state_machine(factory, data):
                         # If it does, and the result is a 'MultipleResult',
                         # then 'print_step' prints a multi-variable assignment.
                         machine._print_step(rule, data_to_print, result)
-                machine.check_invariants(settings)
+                try:
+                    machine.check_invariants(settings)
+                finally:
+                    if print_steps:
+                        machine._print_invariants()
```

Comment:
Instead of adding a new `_print_invariants()` method, I'd rather just add a `report()` line inside `check_invariants()` - just before `result = invar.function(self)`, so that the precondition checks etc. are all handled in one place.

That'll involve duplicating the `print_steps` condition, but I'm OK with that.